torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -165,18 +164,24 @@ def _init_grad_divide_factors(self):
         data_parallel_world_size *= self.mesh_info.shard_mesh_size
         if isinstance(self.mesh_info, HSDPMeshInfo):
             data_parallel_world_size *= self.mesh_info.replicate_mesh_size
+        if self._reduce_dtype == torch.float32:
+            # Only divide after reduction since fp32 has sufficient precision
+            self._grad_predivide_factor: float = 1.0
+            self._grad_postdivide_factor: float = data_parallel_world_size
```

Comment:
I actually keep forgetting why we want to do the "manual" division after reduce_scatter, instead of using `ReduceOp.AVG` when calling reduce_scatter?

Reply:
That is a good point. I tested on torchtrain Llama-7B, and `ReduceOp.AVG` is quite a bit faster I think since we avoid extra memory read/write for the div (as NCCL does it as it is reducing).

I updated the PR with this behavior. I am very interested to understand if we can do this change as well for bf16 reduce-scatter.